api-server/src/main/java/vermillion/http/HttpServerVerticle.java

```diff
@@ -44,1048 +44,1139 @@
+				Object valueObj = attribute.getValue("value");
+				if (!(valueObj instanceof String)) {
+					apiFailure(context, new BadRequestThrowable("Value is not a valid string"));
+					return;
+				}
+
+				String value = attribute.getString("value");
+				attributeQuery = new Queries().getTermQuery();
+				attributeQuery.getJsonObject("term").put("data." + attributeName + ".keyword", value);
+				filterQuery.add(attributeQuery);
+			}
+		}
+
+		baseQuery
+				.put("size", 10000)
+				.getJsonObject("query")
+				.getJsonObject("bool")
+				.put("filter", filterQuery);
+
+		logger.debug(baseQuery.encodePrettily());
+
+		// Trigger regular search function in three cases
+		// 1. When the token provided is null: This is perfectly safe to do at this stage since secure
+		// IDs are checked beforehand
+		// 2. When token is provided but ID is a public ID
+		// 3. When token is provided but ID is a list of public IDs
+		// Don't know why anyone would do 2 & 3, but you never know
+		if ((token == null)
+				|| (resourceIDstr != null && resourceIDstr.endsWith(".public"))
+				|| (resourceIDArray != null
+						&& resourceIDArray.stream()
+								.map(Object::toString)
+								.allMatch(s -> s.endsWith(".public")))) {
+			dbService
+					.rxSearch(baseQuery)
+					.subscribe(
+							result -> response.putHeader("content-type", "application/json").end(result.encode()),
+							t -> apiFailure(context, t));
+		} else {
+			JsonArray requestedIDs = new JsonArray();
+
+			if (resourceIDstr != null) {
+				requestedIDs.add(resourceIDstr);
+			} else {
+				// Get only the secure IDs from the list of all IDs provided
+				requestedIDs =
+						resourceIDArray.stream()
+								.map(Object::toString)
+								.filter(s -> !s.endsWith(".public"))
+								.collect(Collector.of(JsonArray::new, JsonArray::add, JsonArray::add));
+				logger.debug("Requested IDs=" + requestedIDs.encodePrettily());
+			}
+
+			checkAuthorisation(token, READ_SCOPE, requestedIDs)
+					.andThen(Single.defer(() -> dbService.rxSecureSearch(baseQuery, token)))
+					.subscribe(
+							result -> response.putHeader("content-type", "application/json").end(result.encode()),
+							t -> apiFailure(context, t));
+		}
+	}
+
+	// TODO: If Id is provided, reroute to the specific file
+	public void download(RoutingContext context) {
+
+		HttpServerRequest request = context.request();
+
+		// If token is valid for resources apart from secure 'files' then specify list of ids in the
+		// request
+		String token = request.getParam("token");
+		String idParam = request.getParam("id");
+
+		logger.info("token=" + token);
+
+		if (token == null) {
+			apiFailure(context, new BadRequestThrowable("No access token found in request"));
+			return;
+		}
+
+		JsonArray requestedIds = new JsonArray();
+		String basePath = PROVIDER_PATH + "secure/";
+
+		if (idParam != null) {
+			Arrays.asList(idParam.split(",")).forEach(requestedIds::add);
+		}
+
+		for (int i = 0; i < requestedIds.size(); i++) {
```

Comment:
This can be rxified if we could find a way to return from inside lambda. As I understand it, vertx needs an explicit ``return`` to terminate execution when inside an api definition.